Stats package: Add entrypoint class

fgiannar opened the pull request:
Spin off from: https://github.com/Automattic/jetpack/pull/26348
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Adds `Automattic\Jetpack\Stats\Main` entrypoint class for setting hooks and configuration.
* Updates `Automattic\Jetpack\Stats\XMLRPC_Provider` to only add the `jetpack.getBlog` method when the site is connected and the Stats module is active.

#### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
p9dueE-5Rd-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Look at the code and unit tests

fgiannar committed:
Stats package: Add entrypoint class

fgiannar committed:
Stats package: Add entrypoint class

fgiannar committed:
changelog

fgiannar committed:
Merge branch 'trunk' into stats/add-entypoint-class

fgiannar committed:
Add more unit tests

darssen commented on projects/packages/stats/tests/php/test-base.php:
```diff
@@ -81,4 +85,14 @@ protected static function post( $id ) {
+	 * @param array $modules Array with modules slugs.
+	 * @return array
+	 */
+	public static function filter_jetpack_active_modules( $modules ) {
```
This method's name doesn't look representative of what it does.

fgiannar replied:
Updated to `filter_jetpack_active_modules_add_stats`

darssen commented on projects/packages/stats/tests/php/test-main.php:
```diff
@@ -0,0 +1,181 @@
+
+		$this->assertFalse( Stats::jetpack_is_dnt_enabled() );
+	}
+
```
For completeness, I would add a test where the filter is added, but the `HTTP_DNT` header is not present, and check that it returns false.
No filter and no header might be too much :)

fgiannar replied:
Added, thanks! :)

darssen commented on projects/packages/stats/tests/php/test-xmlrpc-provider.php:
```diff
@@ -92,4 +113,14 @@ public function test_get_blog() {
 
 		$this->assertSame( $expected_stats_blog, $stats_blog );
 	}
+
```
This looks to be duplicated from parent class

fgiannar replied:
Nice catch, removed!

darssen reviewed (changes requested):
Nice work :)
I have added a couple of minor comments that can be easily addressed.
Besides that, the only thing I encountered is that the `should_track` method could have some more tests.

fgiannar committed:
Update unit tests

fgiannar commented:
@darssen Many thanks for reviewing, I have addressed all your inline comments :) 

> Besides that, the only thing I encountered is that the should_track method could have some more tests.

Yeah I know, but I focused on explicitly testing the functionalities we added in this PR vs the whole method (that existed anyway in the Jetpack plugin as is). Adding more tests there would be nice indeed but since this is a blocker for the Stats package work I think it's acceptable for now.

darssen commented:
Thanks for addressing them so quickly.
I agree with your reasoning regarding the testing so I am approving :)
